blog/_posts/2018-06-13-scala-213-collections.md

```diff
@@ -0,0 +1,210 @@
+re-implementing a ton of operations. (Some operations, though, still eagerly evaluate
+the collection elements (e.g. `groupBy`) and will be clearly documented.) Another benefit
+is that transformation operations defined outside of the collections (like in the
+[cvogt/scala-extensions](https://github.com/cvogt/scala-extensions) project)
+now work with non-strict collections (such as `View` or `Stream`).
+
+Speaking of non-strict collections, the `View` type has been redesigned and
+views should behave in a more predictable way. Also, `Stream` has been
+deprecated in favor of `LazyList` (see the last section).
+
+## Life Without `CanBuildFrom`
+
+I think the most visible change for end-users is that transformation operations
+don’t use `CanBuildFrom` anymore. I believe this will be quite visible despite our previous
+efforts to *hide* `CanBuildFrom` from the API documentation in the current collections.
+Indeed, if you take a look at the
+[current `List` API](/api/2.12.6/scala/collection/immutable/List.html), the signature
+shown for the `map` operation does not mention `CanBuildFrom`:
+
+![there is no CanBuildFrom](/resources/img/blog/scaladoc-list-map.png)
+
+However, if you use this operation in your code, then your IDE reveals its actual signature:
+
+![what is That](/resources/img/blog/ij-list-map.png)
+
+As you can see, the type signature shown in the API documentation has been “simplified”
+to make it more approachable, but I believe that this is probably introducing more
+confusion to the users. Especially when you look at the
+[`TreeMap[A, B]` API](/api/2.12.6/scala/collection/immutable/TreeMap.html):
+
+![wtf](/resources/img/blog/scaladoc-treemap-map.png)
+
+This type signature makes no sense: the result type can not be `TreeMap[B]` since
+`TreeMap` takes *two* type parameters (the type of keys and the type
+of values). Also, the function `f` actually takes a *key-value pair* as parameter,
+not just a key (as incorrectly indicated by the type `A`).
+
+`CanBuildFrom` was used for good reasons, in particular the type `That` shown
+in the above screenshot was *computed* according to the type of the source
+collection and the type of elements of the new collection. The case of `TreeMap`
+is compelling: in case you transform your key-value pairs into other key-value
+pairs for which the type of keys has an implicit `Ordering` instance, then `map`
+returns a `TreeMap`, but if there is no such `Ordering` instance then the best
+collection type that can be returned is `Map`. And if you transform the key-value
+pairs into something that is not even a pair, then the best collection type
+that can be returned is `Iterable`. These three cases were supported by
+a single operation implementation, and `CanBuildFrom` was used to abstract over
+the various possible return types.
+
+In the new collections we wanted to have simpler type signatures so that we
+can shoulder their actual form in the API documentation and auto-completion
```

Comment:
Is `shoulder` a typo here? I didn't know this verb, it looks like it means "accept the responsibility for something". Does that make sense here?

Reply:
I'd simply say "so that we can show their actual signature in the API, and have IDEs show reasonable completions"